### ISSUE TITLE
Lag/Stutter in Android builds when adding 6 or more ParallaxLayers

### ISSUE THREAD

**Issue by sudiptosarkar** [godotengine/godot#63352]
### Godot version

3.4.4

### System information

Android 12, Redmi Note 10 Pro Max

### Issue description

When adding more than 5 ParallaxLayers, the game stutters to the point of unusability.
It runs fine on my Linux with Nvidia 3080 Ti, but I can see the same issue with my GeForce GT710.

This issue has existed at least since 2020 - https://godotengine.org/qa/84184/how-to-fix-laggy-movement-for-more-than-background-layers-in

### Steps to reproduce

Add 6+ `ParallaxLayer`s under a `ParallaxBackground` with character animation in front, and make the character run in a direction. Set the Motion Scale `x` to increments of 0.1, starting from the furthest layer.

**Comment by Calinou** [comment]
There's a lot of overdraw if you use many ParallaxLayers, so I think this is expected.

**Comment by sudiptosarkar** [comment]
This problem is there on my old laptop where I started the development a few years ago (ATI Radeon Mobility HD5470), and 2015 Macbook Pro.

I had put this issue off for later, hoping that it would be fixed in a later version, but now I'm done with the rest of the stuff in my game, and I need to have at least 6 layers including a foreground parallaxlayer. So I'd really appreciate it if anyone has a workaround to this issue or can point me to how this currently works in the code.

**Comment by Calinou** [comment]
> I had put this issue off for later, hoping that it would be fixed in a later version, but now I'm almost done with the rest of the stuff in my game, and I need to have at least 6 layers including a foreground parallaxlayer. So I'd really appreciate it if anyone has a workaround to this issue or can point me to how this currently works in the code.

Did you try rendering to a lower resolution as I mentioned?

**Comment by sugarsthenewwhite** [comment]
> This problem is there on my old laptop where I started the development a few years ago (ATI Radeon Mobility HD5470), and 2015 Macbook Pro.
> 
> I had put this issue off for later, hoping that it would be fixed in a later version, but now I'm almost done with the rest of the stuff in my game, and I need to have at least 6 layers including a foreground parallaxlayer. So I'd really appreciate it if anyone has a workaround to this issue or can point me to how this currently works in the code.

try using meshed sprites if parallax layers have transparency, which they most likely do.

**Comment by sudiptosarkar** [comment]
> Is the 2015 MacBook Pro a 13" or 15" model? The 15" model has dedicated graphics, whereas the 13" model only has integrated graphics.

It was the 15". It had dedicated Radeon graphics (iirc) and an integrated Intel HD Graphics.

> Did you try rendering to a lower resolution as I mentioned?

Yes, but it doesn't look quite good. I tried 540p (I was running 720p before). I tried it after converting the background sprites to 2D Meshes like @sugarsthenewwhite suggested though, so I didn't face that problem.

> Also, make sure to switch to the GLES2 backend if you haven't done so already. It'll perform better on low-end devices compared to GLES3.

Well, my phone's mid-range... :smiling_face_with_tear:
I have run Asphalt 9 and Real Racing 3 on it without any problems whatsoever. I doubt it's the OpenGL driver.

**Comment by sudiptosarkar** [comment]
> try using meshed sprites if parallax layers have transparency, which they most likely do.

Thank you so much, @sugarsthenewwhite! :pray: 
Your approach worked beautifully, and I was even able to switch up my resolution from 720p to 1080p without the stuttering.

I was then facing a less frequent stuttering. Thanks to the profiler, I figured out it was because of my terrain generation script, which was taking like a 100ms. Good thing we have Threads and Semaphores, so I'm confident I can fix that problem.

So, in conclusion the `ParallaxLayer`s are now working flawlessly.

Now, considering the 2D Mesh feature was added in 3.0 to specifically address this issue (among other transparency rendering issues), I'd leave it up to the maintainers to make a decision on whether to close this issue.

**Comment by sugarsthenewwhite** [comment]
> > try using meshed sprites if parallax layers have transparency, which they most likely do.
> 
> Thank you so much, @sugarsthenewwhite! :pray: 
> Your approach worked beautifully, and I was even able to switch up my resolution from 720p to 1080p without the stuttering.
> 
> I was then facing a less frequent stuttering. Thanks to the profiler, I figured out it was because of my terrain generation script, which was taking like a 100ms. Good thing we have Threads and Semaphores, so I'm confident I can fix that problem.
> 
> So, in conclusion the `ParallaxLayer`s are now working flawlessly.
> 
> Now, considering the 2D Mesh feature was added in 3.0 to specifically address this issue (among other transparency rendering issues), I'd leave it up to the maintainers to make a decision on whether to close this issue.

No problem, these kind of optimization tricks gotta be more common knowledge, i probably wouldn't have known it either if it wasn't some guy on StackOverFlow.
